src/Riok.Mapperly/SyntaxProvider.Roslyn3.9.cs

```diff
@@ -0,0 +1,46 @@
+﻿#if ROSLYN3_9_OR_GREATER && !ROSLYN4_0_OR_GREATER
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Riok.Mapperly;
+
+class SyntaxProvider : ISyntaxContextReceiver
+{
+    public List<ClassDeclarationSyntax> ClassDeclarations { get; } = new();
```

Comment:
I think this can be converted to a private readonly field and an additionally `IReadOnlyCollection` can be exposed to encapsulate the modification of the collection.